### PR TITLE
External project build system

### PR DESCRIPTION
**Pull request by patmarion**

### CODE REVIEW

**Comment by patmarion** [comment]
Add new build system based on cmake external project.  Based on earlier work by @RussTedrake.

Not ready to merge yet, needs further testing.

By default, bot2-vis is disabled, meaning some other libbot libs are also disabled.

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,87 @@
+cmake_minimum_required(VERSION 2.8.5)
+project(libbot)
+
```
Comment (RussTedrake) [inline CMakeLists.txt]: i think you should add the logic which looks for the BUILD folder in here...

File: CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+# set a default build type if none was specified
+if(NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)
+  message(STATUS "Setting build type to 'Release' as none was specified.")
+  set(CMAKE_BUILD_TYPE Release CACHE STRING "Choose the type of build." FORCE)
+  set_property(CACHE CMAKE_BUILD_TYPE PROPERTY STRINGS
+    "Debug" "Release" "MinSizeRel" "RelWithDebInfo")
+endif()
+
+option(WITH_BOT_VIS "Build with bot2-vis" OFF)
```
Comment (RussTedrake) [inline CMakeLists.txt]: maybe add the explanation that it incurs the gtk+ dependency, which causes problems on mac (and reference our issue)?

Comment (RussTedrake) [inline CMakeLists.txt]: wish there was something clever we could do to install_prereqs.sh to capture these two cases, too.

Comment (RussTedrake) [inline CMakeLists.txt]: actually, how about off by default on mac only...?

Comment (RussTedrake) [inline CMakeLists.txt]: even better -- you could check for gtk.

**Comment by patmarion** [comment]
A test we could run in cmake is to look for the existence of gdk/gdkx.h, that will determine whether gtk was compiled with x11 support or not.  It's probably common for people to have homebrew gtk installed, but without x11 support.  So we could set WITH_BOT_VIS = header_exists(gdk/gdkx.h).

But, you know I'm not really a fan of setting defaults based on the environment.  When possible, I think it's better to use a predictable set of defaults, just makes things simpler, build systems can always benefit from being simpler :)

**Comment by RussTedrake** [comment]
i’m willing to be convinced if you really think simpler is better.  but this does seem like it would do the right thing in every case I would imagine — and it would be nice for the text to work by default on systems where it is supported.

**Comment by patmarion** [comment]
One last detail: the libbot build now does some recompile work each time it is invoked.  The way cmake external project works is a little different than tobuild.txt.  CMake will force a reconfigure step if a previous project's install step was executed.  The way libbot's CMakeLists.txt are written, if configure is re-executed then some files are touched (probably lcm generated files) and that causes a recompile.  It's not a ton of time, and can probably be resolved later if we want (by fixing the lcmtype generation rules)

**Pull request merged**